From 196b4cbbcab23ea9f6caaa0dd780ac68dd287048 Mon Sep 17 00:00:00 2001
From: Andrew Kelley <andrew@ziglang.org>
Date: Sun, 7 Sep 2025 17:28:01 -0700
Subject: [PATCH 098/133] frontend: vectors and arrays no longer support
 in-memory coercion
X-Developer-Signature: v=1; a=openpgp-sha256; l=3839; i=xtex@astrafall.org;
 h=from:subject; bh=XOrDtJZwdPUKRjkoYFB9EAz8xdbPYqvP/GkKsXYD6G0=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78rhaGT5Jc3RLPOCJniPcM7FfW6o2I9CRz/ybVM6D7
 a6THn7vmMjCIMbFYCmmyFJk2ODNqpPOL7qsXBZmDisTyBBpkQYGBgYGFga+3MS8UiMdIz1TbUM9
 Q0MdIx0jBi5OAZhqBmVGhl0ryhzO3ll5fpF48q7D/zhYWKTbv97aLL1SamPsPJ+lUy4zMjyLUN5
 2b7/Yr+uCJ8+Iv8j9NfX3o/eu8nzHeycdu3iOl5kdAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

closes #25172

(cherry picked from commit 6673b476855931137119ae92e91fdcb2822e9c4f)
---
 doc/langref.html.in         |  8 +++++++
 doc/langref/test_vector.zig |  2 +-
 src/Sema.zig                | 44 -------------------------------------
 3 files changed, 9 insertions(+), 45 deletions(-)

diff --git a/doc/langref.html.in b/doc/langref.html.in
index bdfcd7f0252c..d978a805d7f6 100644
--- a/doc/langref.html.in
+++ b/doc/langref.html.in
@@ -1969,6 +1969,14 @@ or
       </p>
       {#see_also|@splat|@shuffle|@select|@reduce#}
 
+      {#header_open|Relationship with Arrays#}
+      <p>Vectors and {#link|Arrays#} each have a well-defined <strong>bit layout</strong>
+      and therefore support {#link|@bitCast#} between each other. {#link|Type Coercion#} implicitly peforms
+      {#syntax#}@bitCast{#endsyntax#}.</p>
+      <p>Arrays have well-defined byte layout, but vectors do not, making {#link|@ptrCast#} between
+      them {#link|Illegal Behavior#}.</p>
+      {#header_close#}
+
       {#header_open|Destructuring Vectors#}
       <p>
         Vectors can be destructured:
diff --git a/doc/langref/test_vector.zig b/doc/langref/test_vector.zig
index cd1e860886e1..933a28291001 100644
--- a/doc/langref/test_vector.zig
+++ b/doc/langref/test_vector.zig
@@ -17,7 +17,7 @@ test "Basic vector usage" {
 }
 
 test "Conversion between vectors, arrays, and slices" {
-    // Vectors and fixed-length arrays can be automatically assigned back and forth
+    // Vectors can be coerced to arrays, and vice versa.
     const arr1: [4]f32 = [_]f32{ 1.1, 3.2, 4.5, 5.6 };
     const vec: @Vector(4, f32) = arr1;
     const arr2: [4]f32 = vec;
diff --git a/src/Sema.zig b/src/Sema.zig
index 97fcc9d503d7..5c89060f02df 100644
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -29690,50 +29690,6 @@ pub fn coerceInMemoryAllowed(
         return .ok;
     }
 
-    // Arrays <-> Vectors
-    if ((dest_tag == .vector and src_tag == .array) or
-        (dest_tag == .array and src_tag == .vector))
-    {
-        const dest_len = dest_ty.arrayLen(zcu);
-        const src_len = src_ty.arrayLen(zcu);
-        if (dest_len != src_len) {
-            return .{ .array_len = .{
-                .actual = src_len,
-                .wanted = dest_len,
-            } };
-        }
-
-        const dest_elem_ty = dest_ty.childType(zcu);
-        const src_elem_ty = src_ty.childType(zcu);
-        const child = try sema.coerceInMemoryAllowed(block, dest_elem_ty, src_elem_ty, dest_is_mut, target, dest_src, src_src, null);
-        if (child != .ok) {
-            return .{ .array_elem = .{
-                .child = try child.dupe(sema.arena),
-                .actual = src_elem_ty,
-                .wanted = dest_elem_ty,
-            } };
-        }
-
-        if (dest_tag == .array) {
-            const dest_info = dest_ty.arrayInfo(zcu);
-            if (dest_info.sentinel != null) {
-                return .{ .array_sentinel = .{
-                    .actual = Value.@"unreachable",
-                    .wanted = dest_info.sentinel.?,
-                    .ty = dest_info.elem_type,
-                } };
-            }
-        }
-
-        // The memory layout of @Vector(N, iM) is the same as the integer type i(N*M),
-        // that is to say, the padding bits are not in the same place as the array [N]iM.
-        // If there's no padding, the bitcast is possible.
-        const elem_bit_size = dest_elem_ty.bitSize(zcu);
-        const elem_abi_byte_size = dest_elem_ty.abiSize(zcu);
-        if (elem_abi_byte_size * 8 == elem_bit_size)
-            return .ok;
-    }
-
     // Optionals
     if (dest_tag == .optional and src_tag == .optional) {
         if ((maybe_dest_ptr_ty != null) != (maybe_src_ptr_ty != null)) {
-- 
2.52.0

